[Woo POS] Specialise card reader modal action handlers

joshheald opened the pull request:
<!-- Remember about a good descriptive title. -->

Part of: #12868 
<!-- Id number of the GitHub issue this PR addresses. -->

## Description
<!-- Take the time to write a good summary. Why is it needed? What does it do? When fixing bugs try to avoid just writing “See original issue” – clarify what the problem was and how you’ve fixed it. -->

This PR resolves issues around some of the reused action handlers on connection and payment modals, by making them work even when no viewcontroller is present.

Prior to this PR, existing UIKit action handlers were adapted to be called via [this extension](https://github.com/woocommerce/woocommerce-ios/blob/ae36fbf05ba0a8343e13862a0e14e89739e9083c/WooCommerce/Classes/ViewModels/CardPresentPayments/CardPresentPaymentsModalViewModel.swift#L73) on
``` extension CardPresentPaymentsModalViewModelUIKitActions where Self: CardPresentPaymentsModalViewModelActions```.

This was a naive implementation, as it relied on the handlers working even when passed `nil` instead of a UIViewController.

This PR goes through the existing view controllers and improves this, by providing specific handlers to call the underlying code even when there's no view controller.

It doesn't cover every case yet – we will need to go further with a specialised `alertProvider` for the adapted code to cover some of the views, but it helps us use a lot more of the flows.

## Steps to reproduce
<!-- Step-by-step testing instructions. For new user flows, consider instead stating the goal of the workflow and see if your PR reviewer can accomplish the workflow without specific steps! -->

1. Launch the app using a store eligible for POS
2. Navigate to `Menu > Point of Sale mode`
3. Add a product
4. Tap `Checkout`
5. Tap `Card`
6. Try the buttons that show up – observe that they (mostly) work as expected.

## Testing information
<!-- This is your opportunity to break out individual scenarios that need testing (when necessary) and/or include a checklist for the reviewer to go through. Consider documenting the following from your own completed testing: devices used, alternate workflows, edge cases, affected areas, critical flows, areas not tested, and any remaining unknowns. Provide feedback on this new section of the PR template through Sept 30, 2024 to Apps Quality; additional context here: p91TBi-b8z-p2#comment-12036 -->

I removed the unused `CardPresentModalRetryableError` in this PR. Our retryable errors are all displayed with specific errors, or `CardPresentModalError`, so it didn't make sense to spend time adapting the one we don't use.

Some handlers don't fully work yet – these are noted with TODO comments. I think we will fix them by injecting a different alertProvider. That's probably the MVP solution anyway; it will allow us to have specialised messages for the smaller space available on the POS designs, but it's not needed just yet.

You can simulate `CardPresentModalScanningFailed` by turning off Bluetooth then trying to connect to a real reader.

Some errors are most easily simulated by changing the code – e.g. adding an error at the [start of `checkOrderIsStillEligibleForPayment`](https://github.com/woocommerce/woocommerce-ios/blob/ae36fbf05ba0a8343e13862a0e14e89739e9083c/WooCommerce/Classes/ViewRelated/Orders/Collect%20Payments/CollectOrderPaymentUseCase.swift#L243):
```
return onCheckCompletion(.failure(CollectOrderPaymentUseCaseError.unknownErrorRefreshingOrder))
```

This will result in a `CardPresentModalError` being created, and its new handler implementations used. Previously, the "Back to Order" (needs renamed) button would have failed because there was no view controller. Unfortunately, this still doesn't work, as per the note, but you can at least see the new one is called using a breakpoint.

---
- [x] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

joshheald committed:
12868 Hide receipt actions on adapted success modal

dangermattic commented:
<!--
  0 Errors
  1 Warning: PR is not assigned to a milest...
  0 Messages
  0 Markdowns
-->
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Warning">
          1 Warning
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">PR is not assigned to a milestone.</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_pr-check">
  Generated by :no_entry_sign: <a href="https://danger.systems/">Danger</a>
</p>

joshheald committed:
Merge branch 'issue/12867-collect-payment-on-new-card-present-payment-service' into issue/12868-specialise-action-handlers-where-needed

joshheald committed:
Merge branch 'issue/12867-collect-payment-on-new-card-present-payment-service' into issue/12868-specialise-action-handlers-where-needed

joshheald committed:
Merge branch 'issue/12867-collect-payment-on-new-card-present-payment-service' into issue/12868-specialise-action-handlers-where-needed

joshheald committed:
12868 Add required specialised action handlers

Anywhere we use a ViewController to handle a button tap won’t work when used via the default implementation on `CardPresentPaymentsModalViewModelActions`, because the default implementation calls the UIKit handlers with a nil VC.

This commit adds specialised handlers for each of the affected modals

joshheald committed:
12868 Update handlers with whether they work

joshheald committed:
12868 Remove unused RetryableError

joshheald committed:
12868 Update comment: we have some specific handers

jaclync commented on WooCommerce/Classes/ViewModels/CardPresentPayments/CardPresentModalError.swift:
```diff
@@ -61,6 +61,25 @@ final class CardPresentModalError: CardPresentPaymentsModalViewModel {
+            title: secondaryButtonTitle,
+            actionHandler: { [weak self] in
+                self?.dismissCompletion()
+            })
```
super nit: updated in https://github.com/woocommerce/woocommerce-ios/pull/12946/commits/898910791d2a3c7c3f886ea80fc1dc5dc47cd89d
```suggestion
            actionHandler: dismissCompletion)
```

jaclync committed:
Nit update.

jaclync committed:
Update function return value type to conform to `CardPresentPaymentsModalViewModelActions` protocol.

jaclync commented on WooCommerce/Classes/ViewModels/CardPresentPayments/CardPresentModalTapCard.swift:
```diff
@@ -77,6 +77,15 @@ final class CardPresentModalTapCard: CardPresentPaymentsModalViewModel {
 
+// CardPresentPaymentsModalViewModelActions
+extension CardPresentModalTapCard {
+    var secondaryButtonViewModel: CardPresentPaymentsModalButtonViewModel {
```
While testing this state with breakpoints in `onCancel`, I noticed that `onCancel` and `secondaryButtonViewModel` weren't invoked and thus the cancel CTA was no-op because of the fallback to `didTapSecondaryButton` with nil viewController.

After updating the function signature to return an optional value in https://github.com/woocommerce/woocommerce-ios/pull/12946/commits/9ff33633b634a4c86e0177bb534c44240bc3b63a, `onCancel` is invoked and the cancelation works now.

joshheald replied:
Nice catch! Thanks.

jaclync reviewed (approved):
Nice work! Thanks for going over many view models, I tested all the view cases with manual code changes and it worked as expected in the TODO comments. The only thing I noticed was the missing action handling in `CardPresentModalTapCard` but it was a simple fix. Merging after CI!
